javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -1,75 +1,91 @@
 package java.util
 
-// Ported from Harmony
+// Ported from Scala.js, commit: 00915e8, dated: 2020-09-29
 
 import java.io._
-import java.lang.StringBuilder
-import java.math.BigDecimal
-import java.math.BigInteger
-import java.math.MathContext
-import java.nio.CharBuffer
+import java.lang.{
+  Double => JDouble,
+  Float => JFloat,
+  Boolean => JBoolean,
+  StringBuilder => JStringBuilder
+}
+import java.math.{BigDecimal, BigInteger, MathContext, RoundingMode}
 import java.nio.charset.Charset
-import java.text.DateFormatSymbols
-import java.text.DecimalFormat
-import java.text.DecimalFormatSymbols
-import java.text.NumberFormat
-
-import scala.util.control.Breaks
+import scala.annotation.switch
+import scala.scalanative.regex.Pattern
 
-class Formatter(
-    private var _out: Appendable,
-    private var _locale: Locale
-) extends Closeable
+final class Formatter private (private[this] var dest: Appendable,
+                               formatterLocaleInfo: Formatter.LocaleInfo)
+    extends Closeable
     with Flushable {
 
-  if (_out == null)
-    _out = new StringBuilder()
-
-  private var closed: Boolean = false
-
-  private var lastIOException: IOException = _
+  import Formatter._
+  import Defaults._
+  import Flags._
+
+  /** If `dest == null`, the real content is in `stringOutput`.
+   *
+   *  A real `StringBuilder` may be created lazily if `out()` is called, which
+   *  will then capture the current content of `stringOutput`.
+   *
+   *  This allows to bypass the allocation of the `StringBuilder`, the call
+   *  through `dest.append()` and more importantly the `try..catch`es in the
+   *  common case where the `Formatter` is created without a specific
+   *  destination.
+   */
+  private[this] var stringOutput: String = ""
```

Comment:
I don't think this optimization makes sense in Scala Native. In JavaScript, `j.l.StringBuilder` is implemented on top of string concatenation, and we rely on the JS VM to be efficient about that.

In Native, however, string concatenation is not a primitive. It makes more sense to always create a `j.l.StringBuilder`, since that will reduce code paths.